File: shadow-method/shadow.py
class a():
    def __init__(self):
        print("init")
    def showa(self):
        print("show a")
        self.showb()
    def showb(self):
        print("show b")




class b(a):
    def __init__(self):
        # 继承的时候如果不主动调用就不会初始化
        super(b, self).__init__()
        print("init teacher")
    def showb(self):
        print("show teacher")



if __name__ =="__main__":
    c=b()
    c.showa()
    c.showb()
    # init
    # init teacher
    # show a
    # show teacher
    # show teacher

File: shadow-method/shadow.go
package main

import "fmt"

type People struct{}

func (p People) ShowA() {
	fmt.Println("showA")
	p.ShowB()
}

func (p *People) ShowB() {
	fmt.Println("showB")
}

type Teacher struct {
	People
}

func (t *Teacher) ShowB() {
	fmt.Println("teacher showB")
}

func main() {
	t := Teacher{}
	t.ShowA()
		// 关于继承方法， teacher 生成了 showA 方法，但是传递的参数还是people 所以调用的时候还是调用的  b 的showB的方法
	// 生成的包装方法如下
	//func (t *Teacher) ShowA() {
	//	(*People).ShowA(&(t.People))
	//}
	//showA
	//showB
}
